### PR TITLE
1164 feat tabular new check correlation between features

### PR DESCRIPTION
**Pull request by TheSolY**
#### Reference Issue 
1164

#### What does this implement/fix? 
New data integrity check feature-feature correlation to check the pairwise correlations between features

### CODE REVIEW

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    def run_logic(self, context: Context, dataset_type: str = 'train') -> CheckResult:
+        """
+        Run Check.
+
+        Returns
+        -------
+        CheckResult
+        A heatmap of the pairwise correlations between the features
+        """
+        if dataset_type == 'train':
```
Comment (JKL98ISR) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:50]: there's a function in the context for it, (get dataset by kind or something).

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        full_df.loc[cat_features, cat_features] = cat_cat_corr
+        full_df.loc[num_features, cat_features] = num_cat_corr
+        full_df.loc[cat_features, num_features] = num_cat_corr.transpose()
+
+        # Display
+        fig = px.imshow(full_df)
+
+        return CheckResult(value=full_df, header='Feature-Feature Correlation', display=fig)
+
+    def add_condition_max_number_of_pairs_above(self, threshold: float = 0.9, n_pairs: int = 0):
```
Comment (JKL98ISR) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:84]: maybe add threshold in the name? and maybe keep using greater/less

File: deepchecks/tabular/datasets/classification/adult.py [hunk 3]
```diff
@@ -154,7 +154,7 @@ def load_data(data_format: str = 'Dataset', as_train_test: bool = True) -> \
         tuple if as_train_test = True. Tuple of two objects represents the dataset splitted to train and test sets.
     """
     if not as_train_test:
-        dataset = pd.read_csv(_FULL_DATA_URL)
+        dataset = pd.read_csv(_FULL_DATA_URL, names=_FEATURES + [_target])
```
Comment (JKL98ISR) [inline deepchecks/tabular/datasets/classification/adult.py:157]: what is it doing?

Comment (TheSolY) [reply deepchecks/tabular/datasets/classification/adult.py:157]: fixing a bug - the csv doesn't have a header with the column names

Comment (JKL98ISR) [reply deepchecks/tabular/datasets/classification/adult.py:157]: oh cool

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 1 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:50]: I see it exists but not used in any of the other integrity checks, also requires DatasetKind object instead of string

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 2 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:84]: I like the name but open for discussion

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 1 above)
Comment (JKL98ISR) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:50]: oh maybe we only did it in vision

**Review by JKL98ISR — APPROVED** [review]

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+
+class FeatureFeatureCorrelation(SingleDatasetCheck):
+    """Checks for pairwise correlation between the features.
+
+        Extremely correlated pairs could indicate redundancy and even duplication.
+
+    Parameters
+    ----------
+        n_top_pairs : int , default: 5
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:31]: if this parameter is for display only, it should be called _to_show or _to_display

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        self.n_top_pairs = n_top_pairs
+
+    def run_logic(self, context: Context, dataset_type: str = 'train') -> CheckResult:
+        """
+        Run Check.
+
+        Returns
+        -------
+        CheckResult
+        A heatmap of the pairwise correlations between the features
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:48]: indentation :)

**Comment by TheSolY** [comment]
This is what the output looks like on the adult dataset @nirhutnik 
image

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 5 above)
Comment (nirhutnik) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:48]: Also, the heatmap is less interesting - describe the format of the .value

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        cat_features = dataset.cat_features
+
+        encoded_cat_data = dataset.data.loc[:, cat_features].apply(lambda x: pd.factorize(x)[0])
+
+        # Numerical-numerical correlations
+        num_num_corr = dataset.data.loc[:, num_features].corr(method='spearman')
+
+        # Numerical-categorical correlations
+        num_cat_corr = generalized_corrwith(dataset.data.loc[:, num_features], encoded_cat_data,
+                                            method=correlation_ratio)
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:66]: why use this generalized version when .corr gets Callable as method as well? What's the difference?

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        """ Add condition that all pairwise correlations are less than threshold, except for the diagonal."""
+        def condition(result):
+            results_ge = result.ge(threshold)
+            high_corr_pairs = []
+            for i in results_ge.index:
+                for j in results_ge.columns:
+                    if (i == j) | ((j, i) in high_corr_pairs):
+                        continue
+                    if results_ge.loc[i, j]:
+                        high_corr_pairs.append((i, j))
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:94]: Feels like this could be done more efficiently, or at least cleaner, than a double loop. For the least, can use apply and get series name, I think, but maybe there's a better way to get pairs of (index, column) from data frame (and then mask to take only from True, and later filter out duplicates)

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+**Structure:**
+
+* `How are The Correlations Calculated? <#how-are-the-correlations-calculated>`__
+* `Load Data <#load-data>`__
+* `Run the Check <#run-the-check>`__
+* `Define a Condition <#define-a-condition>`__
+
+How are The Correlations Calculated?
+============================
+This check works with 2 types of features: categorical and numerical, and used a different method to calculate the
```
Comment (nirhutnik) [inline docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]:
```suggestion
This check works with 2 types of features: categorical and numerical, and uses a different method to calculate the
```

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+correlation for each combination of feature types.
```
Comment (nirhutnik) [inline docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]:
```suggestion
correlation for each combination of feature types:
```

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+1. numerical-numerical: Pearson's correlation coefficient
+2. numerical-categorical: Correlation ratio
+3. categorical-categorical: Symmetric Theil's U statistic
```
Comment (nirhutnik) [inline docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]: Add links to articles (Wikipedia is fine) that explains all of those, as hyperlinks

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+# =======
+
+import pandas as pd
+from deepchecks.tabular.datasets.classification import adult
+from deepchecks.tabular.checks.data_integrity import FeatureFeatureCorrelation
+
+#%%
+# Load Data
+# ===============
+
```
Comment (nirhutnik) [inline docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]: Add some explanations - "We will load the adult dataset, which is a dataset of ..."

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+#%%
+# Run the Check
+# ===============================================
+
+check = FeatureFeatureCorrelation()
+check.run(ds)
+
+# For displaying the results in an IDE like PyCharm, you can use the following code:
+# check.run(ds).show()
+# The result will be displayed in a new window.
```
Comment (nirhutnik) [inline docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]: We usually don't supply this comment, not sure it's necessary.

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+Feature Feature Correlation
+***************************
+This notebooks provides an overview for using and understanding the feature-feature correlation check.
+
+This check computes the pairwise correlations between the features, potentially spotting pairs of features that are
+highly correlated.
+
+**Structure:**
+
+* `How are The Correlations Calculated? <#how-are-the-correlations-calculated>`__
```
Comment (nirhutnik) [inline docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]: I'm missing here some explanations on what this means, and what should I do.

**Review by nirhutnik — COMMENTED** [review]
Really good! most of these comments are nitpicking, some are structural or about code design.

File: deepchecks/utils/distribution/preprocessing.py [hunk 14]
```diff
@@ -188,3 +188,22 @@ def preprocess_2_cat_cols_to_same_bins(dist1: Union[np.ndarray, pd.Series], dist
...
+    x: Union[List, np.ndarray, pd.Series]
+        A sequence of a categorical variable values without nulls
+    Returns:
+    --------
+    List[float]
+        Representing the value frequency of x
+    """
+    x_values_counter = Counter(x)
+    total_occurrences = sum(x_values_counter.values())
+    values_probabilities = list(map(lambda n: n / total_occurrences, x_values_counter.values()))
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/utils/distribution/preprocessing.py:208]: x could also be a list or an np array

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py (hunk 12 above)
Comment (TheSolY) [reply docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]: It's a new display and I think we should add it to other examples :)

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+from deepchecks.core import CheckResult, ConditionCategory, ConditionResult
+from deepchecks.tabular import Context, SingleDatasetCheck
+from deepchecks.utils.correlation_methods import correlation_ratio, symmetric_theil_u_correlation
+from deepchecks.utils.dataframes import generalized_corrwith
+
+__all__ = ['FeatureFeatureCorrelation']
+
+
+class FeatureFeatureCorrelation(SingleDatasetCheck):
+    """Checks for pairwise correlation between the features.
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:25]: What other term would you suggest?

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 6 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:66]: .corr is with self, .corrwith is with another dataframe with the same column names, generalized_corrwith is between any 2 dataframes with the same number of rows

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 7 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:94]: I looked for a better way but couldn't find any that was still legible - open for suggestions

File: deepchecks/utils/dataframes.py [hunk 16]
```diff
@@ -126,3 +126,31 @@ def select_from_dataframe(
...
+        Method of correlation. callable with input two 1d ndarrays and returning a float.
+
+    Returns
+    -------
+    DataFrame
+        Pairwise correlations.
+    """
+    corr_results = pd.DataFrame(index=x1.columns, columns=x2.columns)
+    for col2 in x2.columns:
+        corr_results.loc[:, col2] = x1.corrwith(x2[col2], method=method)
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/utils/dataframes.py:155]: great idea

**Comment by nirhutnik** [comment]
@TheSolY About the display - it's really great :) Would only add a short explanation under the heatmap itself, as is done in FeatureLabelCorrelation (but much, much shorter). It should explain that this is a heatmap of correlation between features.

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 2 above)
Comment (nirhutnik) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:78]: Agree with JKL. We should keep our usual naming convention + incomplete sentence seems weird to me

File: deepchecks/utils/distribution/preprocessing.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/utils/distribution/preprocessing.py:208]: Could still check for type. But I agree this is not critical in any way.
Though the fix for "len" will not work for series, I think?

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nirhutnik) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py]: Maybe just simply "Checks for correlation between all possible pairs of features" or something of that sort

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 6 above)
Comment (nirhutnik) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py]: Ohhh ok, thanks :)

File: deepchecks/utils/distribution/preprocessing.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/utils/distribution/preprocessing.py:208]: len works for series

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 7 above)
Comment (nirhutnik) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:88]: Try these suggestions maybe?
https://stackoverflow.com/questions/26854091/getting-index-column-pairs-for-true-elements-of-a-boolean-dataframe-in-pandas

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py (hunk 12 above)
Comment (nirhutnik) [reply docs/source/checks/tabular/data_integrity/plot_feature_feature_correlaion.py]: Ok, sounds good :)

File: deepchecks/utils/dataframes.py [hunk 17]
```diff
@@ -126,3 +126,29 @@ def select_from_dataframe(
...
+        Left data frame to compute correlations.
+    x2: Dataframe
+        Right data frame to compute correlations.
+    method: Callable
+        Method of correlation. callable with input two 1d ndarrays and returning a float.
+
+    Returns
+    -------
+    DataFrame
+        Pairwise correlations, the index match the columns of x1 and the columns match the columns of x2.
```
Comment (nirhutnik) [inline deepchecks/utils/dataframes.py:151]:
```suggestion
        Pairwise correlations, the index matches the columns of x1 and the columns match the columns of x2.
```

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+# Copyright (C) 2021-2022 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""module contains  Feature-Feature Correlation check."""
```
Comment (nirhutnik) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:11]:
```suggestion
"""module contains Feature-Feature Correlation check."""
```

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlation.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/checks/tabular/data_integrity/plot_feature_feature_correlation.py:1]: OK just seen that this happens for education and education num. So just need to make sure the condition catches that (define it such that it will), and say something about it

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+        if num_features:
+            full_df.loc[num_features, num_features] = dataset.data.loc[:, num_features].corr(method='spearman')
+
+        # Categorical-categorical correlations
+        if cat_features:
+            full_df.loc[cat_features, cat_features] = encoded_cat_data.corr(method=symmetric_theil_u_correlation)
+
+        # Numerical-categorical correlations
+        if num_features and cat_features:
+            num_cat_corr = generalized_corrwith(dataset.data.loc[:, num_features], encoded_cat_data,
```
Comment (Nadav-Barak) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:68]: IMO places in which the numeric feature has a null value should be ignored (masked)

**Review by Nadav-Barak — APPROVED** [review]
Looks good!!
Would add 2-3 more tests, specifically one that include null values in different features

File: docs/source/checks/tabular/data_integrity/plot_feature_feature_correlation.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+* `Define a Condition <#define-a-condition>`__
+
+How are The Correlations Calculated?
+============================
+This check works with 2 types of features: categorical and numerical, and uses a different method to calculate the
+correlation for each combination of feature types:
+
+1. numerical-numerical: `Pearson's correlation coefficient <https://en.wikipedia.org/wiki/Pearson_correlation_coefficient>`__
+2. numerical-categorical: `Correlation ratio <https://en.wikipedia.org/wiki/Correlation_ratio>`__
+3. categorical-categorical: `Symmetric Theil's U <https://en.wikipedia.org/wiki/Uncertainty_coefficient>`__
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply docs/source/checks/tabular/data_integrity/plot_feature_feature_correlation.py:24]: I took the formula from Wikipedia link

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+                                                method=correlation_ratio)
+            full_df.loc[num_features, cat_features] = num_cat_corr
+            full_df.loc[cat_features, num_features] = num_cat_corr.transpose()
+
+        # Display
+        fig = px.imshow(full_df)
```
Comment (ItayGabbay) [inline deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:74]: What happens if we have 1000 features? will the plot be readable?
Maybe we should limit to the top X higher correlations?

File: deepchecks/utils/dataframes.py [hunk 23]
```diff
@@ -126,3 +126,29 @@ def select_from_dataframe(
...
+def generalized_corrwith(x1: pd.DataFrame, x2: pd.DataFrame, method: t.Callable):
+    """
+    Compute pairwise correlation.
+
+    Pairwise correlation is computed between columns of one DataFrame with columns of another DataFrame.
+    Pandas' method corrwith only applies when both dataframes have the same column names,
+    this generalized method applies to any two Dataframes with the same number of rows, regardless of the column names.
+
+    Parameters
+    __________
```
Comment (ItayGabbay) [inline deepchecks/utils/dataframes.py:140]: I think it's in the wrong format. should be ---
Don't know if this is github or for real...

Comment (TheSolY) [reply deepchecks/utils/dataframes.py:140]: It's probably Github, in the real code it looks like this
image

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 22 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:74]: I think that the highest are not necessarily the most interesting, and the user should just take this into account and run the check on the partial dataset that they like. Showing only the top could create visual perception bias because the color scheme will be spread on a narrow scale.

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 20 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:68]: the way null values are handled depends on the method

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 22 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py]: @ItayGabbay I've implemented shot_top_n_columns

File: tests/tabular/checks/integrity/feature_feature_correlation_test.py [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+ds = adult.load_data(as_train_test=False)
+
+
+def test_feature_feature_correlation():
+    expected_features = ['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss',
+       'hours-per-week', 'workclass', 'education', 'marital-status',
+       'occupation', 'relationship', 'race', 'sex', 'native-country']
+    result = FeatureFeatureCorrelation().run(ds)
+    assert_that(result.value.index, contains_exactly(*expected_features))
+    assert_that(result.value.columns, contains_exactly(*expected_features))
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply tests/tabular/checks/integrity/feature_feature_correlation_test.py:49]: @nirhutnik @noamzbr  added tests

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+        else:
+            dataset = context.test
+
+        dataset.assert_features()
+        num_features = dataset.numerical_features
+        cat_features = dataset.cat_features
+
+        encoded_cat_data = dataset.data.loc[:, cat_features].apply(lambda x: pd.factorize(x)[0])
+
+        all_features = num_features + cat_features
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:84]: @noamzbr asterisk added to the display

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py]: pairwise means the same think in less words

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+from deepchecks.tabular import Context, SingleDatasetCheck
+from deepchecks.utils.correlation_methods import correlation_ratio, symmetric_theil_u_correlation
+from deepchecks.utils.dataframes import generalized_corrwith
+
+__all__ = ['FeatureFeatureCorrelation']
+
+
+class FeatureFeatureCorrelation(SingleDatasetCheck):
+    """
+    Checks for pairwise correlation between the features.
```
[the remark this answers is not in the document]
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:29]: @noamzbr updated the description

File: deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py (hunk 2 above)
Comment (TheSolY) [reply deepchecks/tabular/checks/data_integrity/feature_feature_correlation.py:117]: @JKL98ISR @nirhutnik suggest concrete names

**Review by nirhutnik — APPROVED** [review]
